compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -1298,21 +1298,38 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
                     if let Some(span) = self.def_span(def_id) {
                         err.span_label(span, &format!("`{}` defined here", path_str));
                     }
-                    let (tail, descr, applicability) = match source {
-                        PathSource::Pat | PathSource::TupleStruct(..) => {
-                            ("", "pattern", Applicability::MachineApplicable)
-                        }
-                        _ => (": val", "literal", Applicability::HasPlaceholders),
+                    let (tail, descr, applicability, old_fields) = match source {
+                        PathSource::Pat => ("", "pattern", Applicability::MachineApplicable, None),
+                        PathSource::TupleStruct(_, args) => (
+                            "",
+                            "pattern",
+                            Applicability::MachineApplicable,
+                            Some(
+                                args.iter()
+                                    .map(|a| {
+                                        self.r.tcx.sess.source_map().span_to_snippet(*a).unwrap()
```

Comment:
Adding a `span_to_snippet` to diagnostic logic doesn't sound like a good idea. Can you instead either
1. Leave field spans in place and add code around them (I think I'd prefer this), or
2. Save field names before hand

?

Reply:
Sorry, maybe I'm missing something but I thought the primary use of `SourceMap::span_to_snippet` was for diagnostic purposes.

I don't think we have access to the old fields, and even if we did that would require stringifying them which `span_to_snippet` already does.

Reply:
Yes, `SourceMap::span_to_snippet` is intended for diagnostics, but we generally try to not use it, leaving it as a last resort when you can't implement the diagnostic otherwise :)

> EDIT: A mere 50 lines above is another span_to_snippet used for error messages.

Yes, but as I said, we'd like to remove most of them, not adding them is still encouraged.

----

You can use [`multipart_suggestion`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic/struct.Diagnostic.html#method.multipart_suggestion) to add field aliases without replacing the whole pattern. I.e. instead of

```
- S(a)
+ S { a: x }
```
you would do
```
S(a)
 - -

S { a: x }
 +++ +++++
```

Spans for adding aliases are just `arg_span.shrink_to_hi()`, `{}` are a bit harder, but something like `path_span.between(first_arg_span)` and `last_arg_span.between(the_whole_thing_span.shrink_to_hi())` should probably work.

Although maybe this will need a slight refactor of the surrounding code, if it's too hard leave a `// FIXME` instead and I'll fix it after this PR is merged.